In [2]:
import pandas as pd
import mysql.connector

conn = mysql.connector.connect(
    host="localhost",
    user="root",
    password="abhay1234",
    database="olist_ecommerce"
)

print("Connected successfully!")

Connected successfully!


In [3]:
orders_df = pd.read_sql("SELECT * FROM orders", conn)
orders_df.head()

C:\Users\abhay\AppData\Local\Temp\ipykernel_19864\1795485520.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  orders_df = pd.read_sql("SELECT * FROM orders", conn)


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29
1,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15
2,000229ec398224ef6ca0657da4fc703e,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05
3,00024acbcdf0a6daa1e931b038114c75,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20
4,00042b26cf59d7ce69dfabb4e55b4fd9,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17


In [5]:
orders_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  str           
 1   customer_id                    99441 non-null  str           
 2   order_status                   99441 non-null  str           
 3   order_purchase_timestamp       99441 non-null  datetime64[us]
 4   order_approved_at              99281 non-null  datetime64[us]
 5   order_delivered_carrier_date   97658 non-null  datetime64[us]
 6   order_delivered_customer_date  96476 non-null  datetime64[us]
 7   order_estimated_delivery_date  99441 non-null  datetime64[us]
dtypes: datetime64[us](5), str(3)
memory usage: 6.1 MB


In [6]:
orders_df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [7]:
orders_df['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [8]:
mismatched = orders_df[(orders_df['order_status'] == 'delivered') & (orders_df['order_delivered_customer_date'].isnull())]
mismatched

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
5168,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
12774,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19
17371,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
17548,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
18041,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
66406,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
89704,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
95490,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16


In [9]:
reverse_mismatch = orders_df[(orders_df['order_status'] != 'delivered') & (orders_df['order_delivered_customer_date'].notnull())]
reverse_mismatch

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
9837,1950d777989f6a877539f53795b4c3c3,1bccb206de9f0f25adc6871a1bcf77b2,canceled,2018-02-19 19:48:52,2018-02-19 20:56:05,2018-02-20 19:57:13,2018-03-21 22:03:51,2018-03-09
17037,2c45c33d2f9cb8ff8b1c86cc28c11c30,de4caa97afa80c8eeac2ff4c8da5b72e,canceled,2016-10-09 15:39:56,2016-10-10 10:40:49,2016-10-14 10:40:50,2016-11-09 14:53:50,2016-12-08
39504,65d1e226dfaeb8cdc42f665422522d14,70fc57eeae292675927697fe03ad3ff5,canceled,2016-10-03 21:01:41,2016-10-04 10:18:57,2016-10-25 12:14:28,2016-11-08 10:58:34,2016-11-25
46160,770d331c84e5b214bd9dc70a10b829d0,6c57e6119369185e575b36712766b0ef,canceled,2016-10-07 14:52:30,2016-10-07 15:07:10,2016-10-11 15:07:11,2016-10-14 15:07:11,2016-11-29
54136,8beb59392e21af5eb9547ae1a9938d06,bf609b5741f71697f65ce3852c5d2623,canceled,2016-10-08 20:17:50,2016-10-09 14:34:30,2016-10-14 22:45:26,2016-10-19 18:47:43,2016-11-30
85048,dabf2b0e35b423f94618bf965fcb7514,5cdec0bb8cbdf53ffc8fdc212cd247c6,canceled,2016-10-09 00:56:52,2016-10-09 13:36:58,2016-10-13 13:36:59,2016-10-16 14:36:59,2016-11-30


In [10]:
# Fix the 8 rows: marked "delivered" but missing a date -> we can't invent a date,
# so we leave the date as-is but flag these for the write-up (can't safely relabel without a date)

# Fix the 6 rows: has a delivery date but status isn't "delivered" -> the date is more
# reliable evidence than the status label, so we correct the status
orders_df.loc[
    (orders_df['order_status'] != 'delivered') & (orders_df['order_delivered_customer_date'].notnull()),
    'order_status'
] = 'delivered'

# Confirm the fix worked
orders_df['order_status'].value_counts()

order_status
delivered      96484
shipped         1107
canceled         619
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [11]:
orders_df['delivery_delay_days'] = (orders_df['order_delivered_customer_date'] - orders_df['order_estimated_delivery_date']).dt.days
orders_df[['order_id', 'order_delivered_customer_date', 'order_estimated_delivery_date', 'delivery_delay_days']].head(10)

,order_id,order_delivered_customer_date,order_estimated_delivery_date,delivery_delay_days
0,00010242fe8c5a6d1ba2dd792cb16214,2017-09-20 23:43:48,2017-09-29,-9.0
1,00018f77f2f0320c557190d7a144bdd3,2017-05-12 16:04:24,2017-05-15,-3.0
2,000229ec398224ef6ca0657da4fc703e,2018-01-22 13:19:16,2018-02-05,-14.0
3,00024acbcdf0a6daa1e931b038114c75,2018-08-14 13:32:39,2018-08-20,-6.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,2017-03-01 16:42:31,2017-03-17,-16.0
5,00048cc3ae777c65dbb7d2a0634bc1ea,2017-05-22 13:44:35,2017-06-06,-15.0
6,00054e8431b9d7675808bcb819fb4a32,2017-12-18 22:03:38,2018-01-04,-17.0
7,000576fe39319847cbb9d288c5617fa6,2018-07-09 14:04:07,2018-07-25,-16.0
8,0005a1a1728c9d785b8e2b08b904576c,2018-03-29 18:17:31,2018-03-29,0.0
9,0005f50442cb953dcd1d21e1fb923495,2018-07-04 17:28:31,2018-07-23,-19.0


In [12]:
# How many days between purchase and actual delivery (total fulfillment time)
orders_df['fulfillment_days'] = (orders_df['order_delivered_customer_date'] - orders_df['order_purchase_timestamp']).dt.days

# A simple yes/no flag: was this order late?
orders_df['is_late'] = orders_df['delivery_delay_days'] > 0

orders_df[['order_id', 'fulfillment_days', 'delivery_delay_days', 'is_late']].head(10)

,order_id,fulfillment_days,delivery_delay_days,is_late
0,00010242fe8c5a6d1ba2dd792cb16214,7.0,-9.0,False
1,00018f77f2f0320c557190d7a144bdd3,16.0,-3.0,False
2,000229ec398224ef6ca0657da4fc703e,7.0,-14.0,False
3,00024acbcdf0a6daa1e931b038114c75,6.0,-6.0,False
4,00042b26cf59d7ce69dfabb4e55b4fd9,25.0,-16.0,False
5,00048cc3ae777c65dbb7d2a0634bc1ea,6.0,-15.0,False
6,00054e8431b9d7675808bcb819fb4a32,8.0,-17.0,False
7,000576fe39319847cbb9d288c5617fa6,5.0,-16.0,False
8,0005a1a1728c9d785b8e2b08b904576c,9.0,0.0,False
9,0005f50442cb953dcd1d21e1fb923495,2.0,-19.0,False


In [13]:
order_items_df = pd.read_sql("SELECT * FROM order_items", conn)

# Combine multiple items into one row per order: total price + total freight
order_totals = order_items_df.groupby('order_id').agg(
    total_price=('price', 'sum'),
    total_freight=('freight_value', 'sum'),
    item_count=('order_item_id', 'count')
).reset_index()

order_totals.head(10)

C:\Users\abhay\AppData\Local\Temp\ipykernel_19864\3385587271.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  order_items_df = pd.read_sql("SELECT * FROM order_items", conn)


,order_id,total_price,total_freight,item_count
0,00010242fe8c5a6d1ba2dd792cb16214,58.90,13.29,1
1,00018f77f2f0320c557190d7a144bdd3,239.90,19.93,1
2,000229ec398224ef6ca0657da4fc703e,199.00,17.87,1
3,00024acbcdf0a6daa1e931b038114c75,12.99,12.79,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,199.90,18.14,1
5,00048cc3ae777c65dbb7d2a0634bc1ea,21.90,12.69,1
6,00054e8431b9d7675808bcb819fb4a32,19.90,11.85,1
7,000576fe39319847cbb9d288c5617fa6,810.00,70.75,1
8,0005a1a1728c9d785b8e2b08b904576c,145.95,11.65,1
9,0005f50442cb953dcd1d21e1fb923495,53.99,11.40,1


In [14]:
orders_full = orders_df.merge(order_totals, on='order_id', how='left')
orders_full.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_delay_days,fulfillment_days,is_late,total_price,total_freight,item_count
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29,-9.0,7.0,False,58.90,13.29,1.0
1,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15,-3.0,16.0,False,239.90,19.93,1.0
2,000229ec398224ef6ca0657da4fc703e,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05,-14.0,7.0,False,199.00,17.87,1.0
3,00024acbcdf0a6daa1e931b038114c75,d4eb9395c8c0431ee92fce09860c5a06,delivered,2018-08-08 10:00:35,2018-08-08 10:10:18,2018-08-10 13:28:00,2018-08-14 13:32:39,2018-08-20,-6.0,6.0,False,12.99,12.79,1.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,58dbd0b2d70206bf40e62cd34e84d795,delivered,2017-02-04 13:57:51,2017-02-04 14:10:13,2017-02-16 09:46:09,2017-03-01 16:42:31,2017-03-17,-16.0,25.0,False,199.90,18.14,1.0
5,00048cc3ae777c65dbb7d2a0634bc1ea,816cbea969fe5b689b39cfc97a506742,delivered,2017-05-15 21:42:34,2017-05-17 03:55:27,2017-05-17 11:05:55,2017-05-22 13:44:35,2017-06-06,-15.0,6.0,False,21.90,12.69,1.0
6,00054e8431b9d7675808bcb819fb4a32,32e2e6ab09e778d99bf2e0ecd4898718,delivered,2017-12-10 11:53:48,2017-12-10 12:10:31,2017-12-12 01:07:48,2017-12-18 22:03:38,2018-01-04,-17.0,8.0,False,19.90,11.85,1.0
7,000576fe39319847cbb9d288c5617fa6,9ed5e522dd9dd85b4af4a077526d8117,delivered,2018-07-04 12:08:27,2018-07-05 16:35:48,2018-07-05 12:15:00,2018-07-09 14:04:07,2018-07-25,-16.0,5.0,False,810.00,70.75,1.0
8,0005a1a1728c9d785b8e2b08b904576c,16150771dfd4776261284213b89c304e,delivered,2018-03-19 18:40:33,2018-03-20 18:35:21,2018-03-28 00:37:42,2018-03-29 18:17:31,2018-03-29,0.0,9.0,False,145.95,11.65,1.0
9,0005f50442cb953dcd1d21e1fb923495,351d3cb2cee3c7fd0af6616c82df21d3,delivered,2018-07-02 13:59:39,2018-07-02 14:10:56,2018-07-03 14:25:00,2018-07-04 17:28:31,2018-07-23,-19.0,2.0,False,53.99,11.40,1.0


In [15]:
orders_full.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
delivery_delay_days              2965
fulfillment_days                 2965
is_late                             0
total_price                       775
total_freight                     775
item_count                        775
dtype: int64

In [18]:
orders_full['is_late'] = (orders_full['delivery_delay_days'] > 0).astype('boolean')
orders_full.loc[orders_full['delivery_delay_days'].isnull(), 'is_late'] = pd.NA
orders_full['is_late'].value_counts(dropna=False)

is_late
False    89941
True      6535
<NA>      2965
Name: count, dtype: Int64

In [19]:
no_items = orders_full[orders_full['total_price'].isnull()]
no_items['order_status'].value_counts()

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [20]:
orders_full.to_csv('orders_cleaned.csv', index=False)

## Data Quality Notes — Phase 3

**Status/date inconsistencies (14 orders, 0.014% of data):**
- 8 orders marked `delivered` with no `order_delivered_customer_date` recorded — left as-is, documented as a gap since no reliable date could be assigned.
- 6 orders had a delivery date recorded but were *not* marked `delivered` (mostly mislabeled as `canceled`) — corrected the status to `delivered`, since the date is stronger evidence than the label.

**Missing delivery dates (2,965 orders, ~3%):**
- These are orders that were canceled, marked unavailable, or otherwise never completed. Missing dates here are expected and meaningful, not an error — they were deliberately left as null rather than filled in.

**Orders with no associated items (775 orders):**
- 767 are `unavailable` or `canceled` orders, which makes sense — no product line was ever finalized.
- 6 are unexplained anomalies (`created`, `invoiced`, `shipped` status with zero items) — flagged but not resolved, since the count is too small to affect the overall analysis.

**Features engineered:**
- `delivery_delay_days` — actual delivery date minus estimated delivery date (negative = early, positive = late).
- `fulfillment_days` — total days from purchase to delivery.
- `is_late` — nullable boolean flag; `True`/`False` where a delivery date exists, `<NA>` where it doesn't (avoids the pandas bug where `NaN > 0` silently evaluates to `False`).
- `total_price`, `total_freight`, `item_count` — aggregated from `order_items` per order.

**Key stat:** Of orders with a known outcome, ~6.8% were delivered later than their estimated date.